File: algorithms/ruby/0104-maximum-depth-of-binary-tree.rb
# frozen_string_literal: true

# Problem: # 104. Maximum Depth of Binary Tree
# URL: https://leetcode.com/problems/maximum-depth-of-binary-tree

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def max_depth(root)
  root ? 1 + [max_depth(root.left), max_depth(root.right)].max : 0
end

File: algorithms/python3/0104-maximum-depth-of-binary-tree.py
# Problem: # 104. Maximum Depth of Binary Tree
# URL: https://leetcode.com/problems/maximum-depth-of-binary-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  def maxDepth(self, root: Optional[TreeNode]) -> int:
    return 1 + max(map(self.maxDepth, (root.left, root.right))) if root else 0